### ISSUE TITLE
InotifyEvents seem to not be closed / garbage collected

### ISSUE THREAD

**Issue by yacn** [gorakhargosh/watchdog#587]
Context: We have a long running service that uses `watchdog` to watch a log file and parse it when it's moved.

Anyways, we've been tracing a memory leak in it with `tracemalloc` and it *appears* like `watchdog` is the culprit, it's the top memory consumer compared to the snapshot taken when starting the app:

```
/opt/app/env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py:0: size=2470 KiB (+2388 KiB), count=47172 (+47157), average=54 B
```

When looking at the tracebacks for this, these are the top memory consuming lines:

```
17427 memory blocks: 851.2 KiB
  File "/opt/app/env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py", line 462
    self._wd = wd
11973 memory blocks: 655.2 KiB
  File "/opt/app/env/lib/python3.6/site-packages/watchdog/observers/inotify_c.py", line 323
    inotify_event = InotifyEvent(wd, mask, cookie, name, src_path)
```

So it seems to me that the issue is the `InotifyEvent` objects that get created are never closed / garbage collected until the `Observer` object stops / joins. 

I've tried issuing a `del event` after we are finished using it but this has no perceptible effect, something else is holding on to a reference to the event.

How could we  "forcibly" close or delete the `InotifyEvent` or `FileSystemEvent` when we're done with it?

**Comment by michelneumann** [comment]
Can confirm. In the project i'm currently working on we use watchdog as well. We also used the `tracemalloc` package to look for the top memory consumers, and using the `InotifyObserver` of watchdog, it was leading the charts with increasing memory usage over time. The scripts were running on a Ubuntu 19.10 machine, Python 3.7.4. The only workaround for us was to explictly use the `PollingObserver` instead of the `Observer` class.

**Comment by BoboTiG** [comment]
Definitely, something may need to be improved. Do you have a clue or even a patch almost ready for that? :crossed_fingers: :)

**Comment by Gandouge78** [comment]
Hey I have the problem on my embedded system.
Each Observer i create on a folder with a lot of activity will increase his memory and provoke my system to kill process.

It's a big issue, does anyone has a solution ?

Polling is not an option for us

**Comment by RoySRC** [comment]
Does this problem still persists?

**Comment by ddsulli1** [comment]
This issue seems to exist in v3.0.0

**Comment by Tantael** [comment]
Confirming, this issue still exists in 3.0.0

**Comment by zhuoyaozzy** [comment]
looks like this issue still exits in the latest version (v6.0)

**Comment by neerava2023** [comment]
Having similar issues, my service has a really fast rate of file changes and the memory usage builds up really quickly.
I have noticed the same areas of concern. 

Also I wanted to add that the `event_queue` is never really cleaned up leading to further build up.
